connector_woocommerce/components/exporter.py

```diff
@@ -0,0 +1,355 @@
+# © 2009 Tech-Receptives Solutions Pvt. Ltd.
+# © 2018 Serpent Consulting Services Pvt. Ltd.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+import psycopg2
+from contextlib import contextmanager
+import odoo
+from odoo import _
+from odoo.addons.component.core import AbstractComponent
+from odoo.addons.connector.exception import (IDMissingInBackend,
+                                             RetryableJobError)
+
+_logger = logging.getLogger(__name__)
+
+"""
+
+Exporters for Woo.
+
+In addition to its export job, an exporter has to:
+
+* check in Woo if the record has been updated more recently than the
+  last sync date and if yes, delay an import
+* call the ``bind`` method of the binder to update the last sync date
+
+"""
+
+
+class WooBaseExporter(AbstractComponent):
+    """ Base exporter for Woo """
+
+    _name = 'woo.base.exporter'
+    _inherit = ['base.exporter', 'base.woo.connector']
+    _usage = 'record.exporter'
+
+    def __init__(self, working_context):
+        super(WooBaseExporter, self).__init__(working_context)
+        self.binding = None
+        self.external_id = None
+
+    def _delay_import(self):
+        """ Schedule an import of the record.
+
+        Adapt in the sub-classes when the model is not imported
+        using ``import_record``.
+        """
+        return
+
+    def _should_import(self):
+        """ Before the export, compare the update date
+        in Woo and the last sync date in Odoo,
+        if the former is more recent, schedule an import
+        to not miss changes done in Woo.
+        """
+        assert self.binding
+        if not self.external_id:
+            return False
+        sync = self.binding.sync_date
+        if not sync:
+            return True
+        result = self.backend_adapter.is_woo_record(self.external_id)
+        return result
+
+    def run(self, binding, *args, **kwargs):
+        """ Run the synchronization
+
+        :param binding: binding record to export
+        """
+        self.binding = binding
+        self.external_id = self.binder.to_external(self.binding)
+        try:
+            should_import = self._should_import()
+        except IDMissingInBackend:
+            self.external_id = None
+            should_import = False
+        if should_import:
+            self._delay_import()
+
+        result = self._run(*args, **kwargs)
+
+        self.binder.bind(self.external_id, self.binding)
+        # Commit so we keep the external ID when there are several
+        # exports (due to dependencies) and one of them fails.
+        # The commit will also release the lock acquired on the binding
+        # record
+        if not odoo.tools.config['test_enable']:
+            binding = binding
+            #self.env.cr.commit()  # noqa
```

Comment:
Why is the commit commented out? It's not on purpose, in the original code ;) (https://github.com/OCA/connector-magento/blob/589b62a5306d74e517c7161e9bc87d3f87172e0f/connector_magento/components/exporter.py#L107)